Add category registry

frenck opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Re-incarnation of https://github.com/home-assistant/core/pull/93498

~~Blocked by #110821 (conflicts in the entity registry)~~

Adds a new registry to Home Assistant: Category registry

This registry holds a set of categories per scope. Simply put: Think of these as scoped set of folders.

Scopes are defined/dictated by the consumer, in our case, the frontend. It can create a scoped category set for different areas of responsibility in the UI. For example, a set of categories for automation, scripts, zones, or scenes.

For the initial PR, the contents for a category is one or more entities.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

frenck committed:
Add category registry

frenck committed:
Add entity registry support

balloob commented on homeassistant/helpers/category_registry.py:
```diff
@@ -0,0 +1,213 @@
+class CategoryEntry:
+    """Category registry entry."""
+
+    category_id: str = field(default_factory=uuid_util.random_uuid_hex)
```
actually ulid is faster. 

```python
from homeassistant.util.ulid import ulid_now
```

balloob commented on homeassistant/components/config/entity_registry.py:
```diff
@@ -152,6 +152,15 @@ def websocket_get_entities(
+        # Therefore, passing in a category ID will either add or move
+        # the entity to that specific category. Passing in None will
+        # remove the entity from the category.
+        vol.Optional("categories"): cv.schema_with_slug_keys(vol.Any(str, None)),
```
How can an entity be in different categories? Shouldn't it only be able to be in the category for the scope that is the same as the entity domain?

frenck replied:
Depends on the view/dashboard.

From the entities dashboard, you could have a different categorization compared to your automations or zones dashboard.

The original PR (#93498) was based on a `domain`, but we changed it to `scope` as it doesn't have to be a domain-limited view (like, for example, an entities dashboard or whatever else).

MartinHjelmare commented on homeassistant/components/config/entity_registry.py:
```diff
@@ -152,6 +152,15 @@ def websocket_get_entities(
+        # If passed in, we update/adjust only the provided scope(s).
+        # Other category scopes in the entity, are left as is.
+        #
+        # Categories can only be in 1 category ID per scope at a time.
```
```suggestion
        # Categorized items such as entities
        # can only be in 1 category ID per scope at a time.
```

balloob commented on homeassistant/helpers/entity_registry.py:
```diff
@@ -161,6 +161,7 @@ class RegistryEntry:
     previous_unique_id: str | None = attr.ib(default=None)
     aliases: set[str] = attr.ib(factory=set)
     area_id: str | None = attr.ib(default=None)
+    categories: dict[str, str] = attr.ib(factory=dict)
```
Same question as above: this can just be a string ID to the category that belongs to the domain ?

balloob commented on homeassistant/helpers/entity_registry.py:
```diff
@@ -1224,6 +1237,15 @@ def _data_to_save(self) -> dict[str, Any]:
+    def async_clear_category_id(self, scope: str, category_id: str) -> None:
+        """Clear category id from registry entries."""
+        for entity_id, entry in self.entities.items():
+            if scope in entry.categories and category_id == entry.categories[scope]:
```
A `entry.categories.get(scope)` should be faster as it only needs to lookup the `scope` key once in the dictionary.

frenck replied:
Addressed in [deb45eb](https://github.com/home-assistant/core/pull/110897/commits/deb45eb3cbb1137b50322a63073bcbb6eb20e44a)

balloob commented on homeassistant/helpers/entity_registry.py:
```diff
@@ -1313,6 +1335,18 @@ def async_entries_for_label(
+    return [
+        entry
+        for entry in registry.entities.values()
+        if scope in entry.categories and entry.categories[scope] == category_id
```
Same

frenck replied:
Addressed in [deb45eb](https://github.com/home-assistant/core/pull/110897/commits/deb45eb3cbb1137b50322a63073bcbb6eb20e44a)

frenck committed:
Update homeassistant/components/config/entity_registry.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

frenck committed:
Use ulid instead

frenck committed:
Add tests for adding same name in different scopes

frenck commented on tests/components/config/test_category_registry.py:
```diff
@@ -0,0 +1,310 @@
+"""Test category registry API."""
+import pytest
```
Added to both, done in [d9c7c92](https://github.com/home-assistant/core/pull/110897/commits/d9c7c9240b83f64d962b814cf6a5324cf526e2e0)

frenck committed:
Handle keyerror on update

frenck commented on homeassistant/components/config/category_registry.py:
```diff
@@ -0,0 +1,132 @@
+    data.pop("id")
+
+    try:
+        entry = category_registry.async_update(**data)
```
Nice catch! Addressed in [a42423e](https://github.com/home-assistant/core/pull/110897/commits/a42423eeef1c2f042d1e704a0b1551371a01481f).

frenck committed:
Lookup tweak

MartinHjelmare commented on homeassistant/components/config/category_registry.py:
```diff
@@ -0,0 +1,134 @@
+    except ValueError as err:
+        connection.send_error(msg["id"], "invalid_info", str(err))
+    except KeyError:
+        connection.send_error(msg["id"], "invalid_info", "Category ID doesn't exist")
```
It could also be the scope that is missing. The error message isn't wrong, as the category entry for that category id wouldn't exist in that case either, but it's confusing if we also don't mention that the scope is missing.

frenck replied:
Well, this is a safeguard, which should (technically not happen). If the category exists, the scope would, too (as a category is nested in the scope).

From the end-user, they can create categories, they can't create scopes. So, from that perspective, I think the current messaging is better in that regard.

frenck committed:
Omit categories from entity registry snapshots

bdraco commented on homeassistant/bootstrap.py:
```diff
@@ -342,6 +343,7 @@ def _cache_uname_processor() -> None:
     template.async_setup(hass)
     await asyncio.gather(
         create_eager_task(area_registry.async_load(hass)),
+        create_eager_task(category_registry.async_load(hass)),
```
Side note: we probably need a way to group the loads in the executor here

bdraco commented on homeassistant/helpers/category_registry.py:
```diff
@@ -0,0 +1,213 @@
+    name: str
+
+
+class CategoryRegistry:
```
There was a base class for registries added a while back that could be used here

frenck replied:
"a while back" 

![CleanShot 2024-03-14 at 20 51 34@2x](https://github.com/home-assistant/core/assets/195327/ed65b5b3-72d6-42a6-8755-40444990bdd5)

🤣

frenck replied:
done in [ef197e3](https://github.com/home-assistant/core/pull/110897/commits/ef197e3338ae4e5015268486bd6e6cf771d6aee2)

bdraco replied:
It felt longer.  😉

frenck committed:
Use base registry

frenck committed:
Update snapshots

MartinHjelmare commented on tests/components/enphase_envoy/snapshots/test_diagnostics.ambr:
```diff
@@ -1637,12 +1701,14 @@
               'domain': 'sensor',
               'entity_category': None,
-              'entity_id': 'sensor.envoy_<<envoyserial>>_lifetime_net_energy_consumption',
+              'entity_id': 'sensor.envoy_<<envoyserial>>_energy',
```
The items besides categories should probably not change.

frenck replied:
Hmm seems like I forgot to update the translations before generating these 🤦

frenck replied:
fixed in [51f8417](https://github.com/home-assistant/core/pull/110897/commits/51f841770180d7f08dda3ccbdffd6a5fc8d28f94)

MartinHjelmare commented:
There's a merge conflict.

frenck commented:
Aah caused by https://github.com/home-assistant/core/pull/113462, which was created and merged during the night. Ok, will resolve the conflict.

frenck committed:
Update snapshots

frenck commented:
Rebased to deal with a merge conflict, updated enphase envoy snapshot in [51f8417](https://github.com/home-assistant/core/pull/110897/commits/51f841770180d7f08dda3ccbdffd6a5fc8d28f94).

frenck commented:
Thanks, @MartinHjelmare 👍 

../Frenck
